psyneulink/library/compositions/pytorchcomponents.py

```diff
@@ -0,0 +1,209 @@
+                            torch.min(input=(x - bias), other=torch.tensor([0], device=device).double()) * leak)
+
+    else:
+        raise Exception(f"Function {function} is not currently supported in AutodiffCompositions!")
+
+class PytorchMechanismWrapper():
+    """
+    An interpretation of a mechanism as an equivalent pytorch object
+    """
+    def __init__(self, mechanism, component_idx, device, context=None):
+        self._mechanism = mechanism
+        self._idx = component_idx
+        self._context = context
+
+        self.function = pytorch_function_creator(mechanism.function, device, context)
+        self._context = context
+        self.value = None
+        self._llvm_value = None
+        self.afferents = []
+        self.efferents = []
+
+        self._target_mechanism = None
+
+    def add_efferent(self, efferent):
+        if not efferent in self.efferents:
+            self.efferents.append(efferent)
+
+    def add_afferent(self, afferent):
+        if not afferent in self.afferents:
+            self.afferents.append(afferent)
+
+
+    def collate_afferents(self):
+        """
+        Returns weight-multiplied sum of all afferent projections
+        """
+        return sum((proj.execute(proj.sender.value) for proj in self.afferents))
+
+    def execute(self, variable):
+        self.value = self.function(variable)
+
+        return self.value
+
+    def _gen_execute_llvm(self, ctx, builder, state, params, mech_input, data):
+        mech_func = ctx.import_llvm_function(self._mechanism)
+
+        mech_param = builder.gep(params, [ctx.int32_ty(0),
+                                          ctx.int32_ty(0),
+                                          ctx.int32_ty(self._idx)])
+
+        mech_state = builder.gep(state, [ctx.int32_ty(0),
+                                         ctx.int32_ty(0),
+                                         ctx.int32_ty(self._idx)])
+
+        mech_output = builder.gep(data, [ctx.int32_ty(0),
+                                         ctx.int32_ty(0),
+                                         ctx.int32_ty(self._idx)])
+
+        builder.call(mech_func, [mech_param,
+                                 mech_state,
+                                 mech_input,
+                                 mech_output])
+
+        pnlvm.helpers.printf_float_array(builder, builder.gep(mech_output, [ctx.int32_ty(0), ctx.int32_ty(0)]), prefix=f"{self} output:\n", override_debug=False)
+
+        return mech_output
+
+    def log_value(self):
+        if self._mechanism.parameters.value.log_condition != LogCondition.OFF:
+            detached_value = self.value.detach().cpu().numpy()
+            self._mechanism.output_port.parameters.value._set(detached_value, self._context)
+            self._mechanism.parameters.value._set(detached_value, self._context)
+
+    def _gen_execute_derivative_func_llvm(self, ctx, builder, state, params, data, arg_in):
+        # HACK: pack input in bigger struct
```

Comment:
why is this hack needed? it looks like it tries to match whatever input shape to that of the mechanism function, but uses custom array instead of the actual input shape.

Reply:
is there any way to construct the expected type of the function?
I tried digging around the `_parse_input_ports` method in Mechanism, but was confused as to what the actual structure should be.

Reply:
you can import the function and check `args[0].type` if there needs to be a type conversion.
do you know why this happens? sounds like another example of shape mismatch. do python shapes match?

Reply:
the function expects a 2d array as input, but I have a vector.

I suspect that this is due to input ports